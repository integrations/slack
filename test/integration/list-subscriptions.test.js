const nock = require('nock');
const request = require('supertest');

const { probot, slackbot, models } = require('.');
const fixtures = require('../fixtures');

const { SlackWorkspace, Installation, Subscription } = models;

describe('Integration: subscription list', () => {
  beforeEach(async () => {
    const workspace = await SlackWorkspace.create({
      accessToken: 'secret',
      slackId: 1,
    });

    const installation = await Installation.create({
      githubId: 1,
      ownerId: 1,
    });

    await Subscription.create({
      channelId: 'C2147483705',
      githubId: 1,
      installationId: installation.id,
      slackWorkspaceId: workspace.id,
      type: 'repo',
    });

    await Subscription.create({
      channelId: 'C2147483705',
      githubId: 2,
      installationId: installation.id,
      slackWorkspaceId: workspace.id,
      type: 'repo',
    });

    await Subscription.create({
      channelId: 'C2147483705',
      githubId: 3,
      installationId: installation.id,
      slackWorkspaceId: workspace.id,
      type: 'account',
    });
  });

  test('works for /github subscribe list', async () => {
    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
      html_url: 'https://github.com/atom/atom',
    });
    nock('https://api.github.com').get('/repositories/2').reply(200, {
      full_name: 'kubernetes/kubernetes',
      html_url: 'https://github.com/kubernetes/kubernetes',
    });
    nock('https://api.github.com').get('/user/3').reply(200, {
      login: 'Microsoft',
      html_url: 'https://github.com/Microsoft',
    });
    const command = fixtures.slack.command({
      text: 'subscribe list',
    });

    await request(probot.server).post('/slack/command').send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });

  test('when a repository and an account has been deleted', async () => {
    probot.logger.level('fatal');

    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
      html_url: 'https://github.com/atom/atom',
    });
    nock('https://api.github.com').get('/repositories/2').reply(404, {});
    nock('https://api.github.com').get('/user/3').reply(404, {
      login: 'Microsoft',
      html_url: 'https://github.com/Microsoft',
    });

    const command = fixtures.slack.command({
      text: 'subscribe list',
    });

    await request(probot.server).post('/slack/command').use(slackbot).send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });

  test('when some other error occurs', async () => {
    probot.logger.level('fatal');

    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
      html_url: 'https://github.com/atom/atom',
    });
    nock('https://api.github.com').get('/repositories/2').reply(500, {});

    const command = fixtures.slack.command({
      text: 'subscribe list',
    });

    await request(probot.server).post('/slack/command').use(slackbot).send(command)
      .expect(200)
      .expect(/ephemeral/)
      .expect(/we had trouble with your request/);
  });

  test('works for /github subscribe', async () => {
    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
      html_url: 'https://github.com/atom/atom',
    });
    nock('https://api.github.com').get('/repositories/2').reply(200, {
      full_name: 'kubernetes/kubernetes',
      html_url: 'https://github.com/kubernetes/kubernetes',
    });
    nock('https://api.github.com').get('/user/3').reply(200, {
      login: 'Microsoft',
      html_url: 'https://github.com/Microsoft',
    });
    const command = fixtures.slack.command({
      text: 'subscribe',
    });

    await request(probot.server).post('/slack/command').use(slackbot).send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });
});
