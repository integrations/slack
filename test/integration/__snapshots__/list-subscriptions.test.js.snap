// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: subscription list when a repository and an account has been deleted 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "fallback": "Subscribed to 1 repository",
      "text": "<https://github.com/atom/atom|atom/atom>",
      "title": "Subscribed to the following repositories",
    },
  ],
  "response_type": "in_channel",
}
`;

exports[`Integration: subscription list when a repository and an account has been deleted 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: subscription list when a repository and an account has been deleted 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "subscribe list",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: subscription list when some other error occurs 1`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: subscription list when some other error occurs 2`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "subscribe list",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: subscription list works for /github subscribe 1`] = `
Object {
  "attachments": Array [
    Object {
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "Subscribe to notifications for a repository:
\`/github subscribe owner/repository\`",
    },
    Object {
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "Unsubscribe from notifications for a repository:
\`/github unsubscribe owner/repository\`",
    },
    Object {
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "Subscribe to notifications for all repositories in an organization:
\`/github subscribe owner\`",
    },
    Object {
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "Unsubscribe from notifications for an organization:
\`/github unsubscribe owner\`",
    },
    Object {
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "List all active subscriptions in a channel:
\`/github subscribe list\`",
    },
    Object {
      "color": "",
      "footer": "<https://github.com/integrations/slack#readme|Learn More> â€” <https://github.com/contact?form%5Bsubject%5D=GitHub%2BSlack%20integration|Contact Support>",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "text": "",
    },
    Object {
      "color": "#24292f",
      "fallback": "Subscribed to 2 repositories",
      "text": "<https://github.com/atom/atom|atom/atom>
<https://github.com/kubernetes/kubernetes|kubernetes/kubernetes>",
      "title": "Subscribed to the following repositories",
    },
  ],
  "response_type": "ephemeral",
  "text": ":wave: Need some help with \`/github\`?",
}
`;

exports[`Integration: subscription list works for /github subscribe 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: subscription list works for /github subscribe 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "subscribe",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: subscription list works for /github subscribe list 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "fallback": "Subscribed to 2 repositories",
      "text": "<https://github.com/atom/atom|atom/atom>
<https://github.com/kubernetes/kubernetes|kubernetes/kubernetes>",
      "title": "Subscribed to the following repositories",
    },
    Object {
      "color": "#24292f",
      "fallback": "Subscribed to 1 account",
      "text": "<https://github.com/Microsoft|Microsoft>",
      "title": "Subscribed to the following accounts",
    },
  ],
  "response_type": "in_channel",
}
`;

exports[`Integration: subscription list works for /github subscribe list 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: subscription list works for /github subscribe list 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "subscribe list",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;
